src/transformers/models/flava/modeling_flava.py

```diff
@@ -0,0 +1,2101 @@
+    generate image tokens for an image based on DALL-E's vocab. To be used to generate labels for MIM. Use
+    `get_codebook_indices` to get image tokens for an image.
+    """,
+    FLAVA_START_DOCSTRING.format(config="FLAVACodebookConfig"),
+)
+class FLAVACodebook(FLAVAPreTrainedModel):
+    base_model_prefix = ""
+    config_class = FLAVACodebookConfig
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+
+    def __init__(
+        self,
+        config: FLAVACodebookConfig,
+        **kwargs: Any,
+    ):
+        super().__init__(config)
+
+        self.config = config
+        self.num_groups = config.num_groups
+        self.input_channels = config.input_channels
+        self.num_blocks_per_group = config.num_blocks_per_group
+        self.hidden_size = config.hidden_size
+        self.vocab_size = config.vocab_size
+
+        num_layers = self.num_groups * self.num_blocks_per_group
+        output_conv = _build_codebook_conv2d(8 * self.hidden_size, self.vocab_size, kernel_size=1)
+
+        self.blocks = nn.Sequential(
+            OrderedDict(
+                [
+                    ("input", _build_codebook_conv2d(self.input_channels, 1 * self.hidden_size, kernel_size=7)),
+                    (
+                        "group_1",
+                        self._create_group(
+                            num_layers, self.num_blocks_per_group, 1 * self.hidden_size, 1 * self.hidden_size
+                        ),
+                    ),
+                    (
+                        "group_2",
+                        self._create_group(
+                            num_layers, self.num_blocks_per_group, 1 * self.hidden_size, 2 * self.hidden_size
+                        ),
+                    ),
+                    (
+                        "group_3",
+                        self._create_group(
+                            num_layers, self.num_blocks_per_group, 2 * self.hidden_size, 4 * self.hidden_size
+                        ),
+                    ),
+                    (
+                        "group_4",
+                        self._create_group(
+                            num_layers,
+                            self.num_blocks_per_group,
+                            4 * self.hidden_size,
+                            8 * self.hidden_size,
+                            use_pool=False,
+                        ),
+                    ),
+                    (
+                        "output",
+                        nn.Sequential(OrderedDict([("relu", nn.ReLU()), ("conv", output_conv)])),
+                    ),
+                ]
+            )
+        )
+        self.post_init()
+
+        if self.config.freeze:
+            self._freeze()
+
+    def _freeze(self):
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def _create_group(
```

Comment:
Refactoring model parts creation in a function like this goes against the general philosohpy of Transformers. It's okay to repeat some code several times in the init.